File: line_graph.R
# This script creates a line graph using (a subset of) 'census_data.csv'.

# A description of 'census_data.csv' can be found in 'census_data_scrape_and_process.R'.
# To avoid an overwhelming plot, I will be using a subset of 'census_data.csv' consisting of
# all observations for Pennsylvania and Illinois.

# Load packages.
library(data.table)
library(reshape2)
library(ggplot2)
library(ggthemes)

# Load data.
census_data = data.table(read.csv('census_data.csv', 
                                  header = T, 
                                  stringsAsFactors = F, 
                                  check.names = F))

#####
# SUBSET/PROCESS DATA.

  # Remove unneeded column.
  census_data$`Area (sq. miles)` = NULL

  # Subset data.
  census_subset = census_data[Name == 'Pennsylvania' | Name == 'Illinois']
  
  # Divide population values by 1000000 for easier graph viewing.
  census_subset$`1960` = census_subset$`1960` / 1000000
  census_subset$`1970` = census_subset$`1970` / 1000000
  census_subset$`1980` = census_subset$`1980` / 1000000
  census_subset$`1990` = census_subset$`1990` / 1000000
  census_subset$`2000` = census_subset$`2000` / 1000000
  census_subset$`2010` = census_subset$`2010` / 1000000
  
  # Reshape data for line graph.
    
    # Transpose 'census_subset'.
    census_subset_transpose = data.table(t(census_subset), 
                                         keep.rownames = T)
    
    # Set column names for 'census_subset_transpose'.
    setnames(census_subset_transpose, 
             names(census_subset_transpose), 
             c('Year', 'Illinois', 'Pennsylvania'))
    
    # Remove old, transposed header row.
    census_subset_transpose = census_subset_transpose[Year != 'Name']
    
    # Melt data into convenient form for line graph.
    census_subset_transpose = melt(census_subset_transpose, 
                                   id = 'Year')
    
    # Change column names for melted 'census_subset_transpose'.
    setnames(census_subset_transpose, 
             names(census_subset_transpose), 
             c('Year', 'State', 'Population'))
    
    # Change 'census_subset_transpose$Population' to 'numeric'.
    census_subset_transpose$Population = as.numeric(census_subset_transpose$Population)

#####  
# Plot data.
ggplot(data = census_subset_transpose, 
       aes(x = Year, 
           y = Population, 
           group = State, 
           color = State)) + 
  geom_line() + 
  geom_point() + 
  ylim(c(10, 13.05)) + 
  theme_economist() +
  theme(axis.text.x = element_text(color = 'black', 
                                   size = 16), 
        axis.text.y = element_text(color = 'black', 
                                   size = 16), 
        axis.title.x = element_text(face = 'bold', 
                                    size = 16),
        axis.title.y = element_text(face = 'bold',
                                    size = 16, 
                                    vjust = 1),
        axis.ticks = element_blank(),
        plot.title = element_text(face = 'bold', 
                                  size = 17, 
                                  hjust = 0.35), 
        legend.title = element_text(size = 17),
        legend.text = element_text(size = 16)) +
  annotate('text', 
           x = 0.91, 
           y = 10.3, 
           label = round(census_subset_transpose[State == 'Illinois' & Year == '1960']$Population, 2), 
           size = 5.5) +
  annotate('text',
           x = 1, 
           y = 11.55, 
           label = round(census_subset_transpose[State == 'Pennsylvania' & Year == '1960']$Population, 2), 
           size = 5.5) + 
  annotate('text', 
           x = 6, 
           y = 12.55, 
           label = round(census_subset_transpose[State == 'Pennsylvania' & Year == '2010']$Population, 2), 
           size = 5.5) +
  annotate('text', 
           x = 6, 
           y = 13.02, 
           label = round(census_subset_transpose[State == 'Illinois' & Year == '2010']$Population, 2), 
           size = 5.5) +
  ggtitle('Population (in milions) Over Time for Selected States') +
  xlab('Year') +
  ylab('Population (in millions)')

File: line_graph.py
# This script creates a line graph using (a subset of) 'census_data.csv'.

# A description of 'census_data.csv' can be found in 'census_data_scrape_and_process.R'.
# To avoid an overwhelming plot, I will be using a subset of 'census_data.csv' consisting of
# all observations for Pennsylvania and Illinois.

# Import modules.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import style

# Set style.
style.use('ggplot')

# Load data.
census_data = pd.read_csv('census_data.csv',
                          encoding = 'latin-1')

# SUBSET/PROCESS DATA.

# Divide population values by 1000000 for easier graph viewing.
census_data['1960'] = census_data['1960'] / 1000000
census_data['1970'] = census_data['1970'] / 1000000
census_data['1980'] = census_data['1980'] / 1000000
census_data['1990'] = census_data['1990'] / 1000000
census_data['2000'] = census_data['2000'] / 1000000
census_data['2010'] = census_data['2010'] / 1000000

# Subset data.
census_subset = census_data[census_data['Name'].isin(['Pennsylvania', 'Illinois'])][['Name', '1960', '1970', '1980', '1990', '2000', '2010']]

# Reset 'census_subset' index.
census_subset.reset_index(drop = True,
                          inplace = True)

# Melt 'census_subset' to allow line graph and rename columns.
census_subset = pd.melt(census_subset,
                        id_vars = 'Name')
census_subset.columns = ['State', 'Year', 'Population']

# PLOT DATA.

# Create figure and subplot.
fig = plt.figure()
ax1 = fig.add_subplot(1, 1, 1)

# Add line data.
ax1.plot(census_subset['Year'].unique(),
         census_subset[census_subset['State'] == 'Pennsylvania']['Population'],
         color = 'b',
         label = 'Pennsylvania')
ax1.plot(census_subset['Year'].unique(),
         census_subset[census_subset['State'] == 'Illinois']['Population'],
         color = 'r',
         label = 'Illinois')

# Add points too.
ax1.scatter(census_subset['Year'].unique().tolist(),
            census_subset[census_subset['State'] == 'Pennsylvania']['Population'],
            color = 'b',
            label = None)
ax1.scatter(census_subset['Year'].unique().tolist(),
            census_subset[census_subset['State'] == 'Illinois']['Population'],
            color = 'r',
            label = None)

# Add plot labels.
ax1.annotate(round(census_subset[(census_subset['State'] == 'Illinois') & (census_subset['Year'] == '1960')]['Population'].values[0], 2),
             xy = (1957, 9.95),
             xytext = (1958, 9.95),
             fontsize = 12)
ax1.annotate(round(census_subset[(census_subset['State'] == 'Pennsylvania') & (census_subset['Year'] == '1960')]['Population'].values[0], 2),
             xy = (1957, 11.15),
             xytext = (1958, 11.185),
             fontsize = 12)
ax1.annotate(round(census_subset[(census_subset['State'] == 'Illinois') & (census_subset['Year'] == '2010')]['Population'].values[0], 2),
             xy = (2008, 12.95),
             xytext = (2008, 12.9),
             fontsize = 12)
ax1.annotate(round(census_subset[(census_subset['State'] == 'Pennsylvania') & (census_subset['Year'] == '2010')]['Population'].values[0], 2),
             xy = (2008, 12.35),
             xytext = (2008.5, 12.55),
             fontsize = 12)

# Set x-axis and y-axis limits.
plt.ylim([9.9, 13.1])
plt.xlim([1955, 2015])

# For first subplot x-axis and y-axis: remove ticks, change tick labels to black and pick size.
ax1.tick_params(axis = 'both',
                length = 0,
                colors = 'black',
                labelsize = 13)

# Remove first and last x-axis label (not needed).
ax1.xaxis.get_major_ticks()[0].label1.set_visible(False)
ax1.xaxis.get_major_ticks()[-1].label1.set_visible(False)

# Adjust plot title spacing.
ax1.title.set_position([0.5, 1.02])

# Set plot and axes titles.
plt.title('Population (in millions) Over Time for 2 States',
          fontweight = 'bold')
plt.xlabel('Year',
           fontweight = 'bold',
           fontsize = 14,
           color = 'black')
plt.ylabel('Population (in millions)',
           fontweight = 'bold',
           fontsize = 14,
           color = 'black')

# Add legend.
legend = ax1.legend(title = 'State',
           loc = 'lower right',
                    fontsize = 15)
plt.setp(legend.get_title(), fontsize = 16)

# Adjust plot margins.
plt.subplots_adjust(left = 0.13,
                    bottom = 0.1,
                    right = 0.98,
                    top = 0.93)

# Show plot.
plt.show()
